drivers/wifi/infineon/airoc_wifi.c

```diff
@@ -0,0 +1,652 @@
+
+		case WLC_E_DEAUTH_IND:
+		case WLC_E_DISASSOC_IND:
+			net_if_dormant_on(airoc_wifi_iface);
+			break;
+
+		default:
+			break;
+		}
+	}
+}
+
+static void airoc_mgmt_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+	struct airoc_wifi_data *data = dev->data;
+	struct ethernet_context *eth_ctx = net_if_l2_data(iface);
+
+	eth_ctx->eth_if_type = L2_ETH_IF_TYPE_WIFI;
+	data->iface = iface;
+	airoc_wifi_iface = iface;
+
+	/* Read WLAN MAC Address */
+	(void)whd_wifi_get_mac_address(airoc_if, &airoc_if->mac_addr);
+	(void)memcpy(&data->mac_addr, &airoc_if->mac_addr, sizeof(airoc_if->mac_addr));
+
+	/* Assign link local address. */
+	(void)net_if_set_link_addr(iface, data->mac_addr, 6, NET_LINK_ETHERNET);
+
+	(void)ethernet_init(iface);
+
+	/* Not currently connected to a network */
+	(void)net_if_dormant_on(iface);
+
+	/* L1 network layer (physical layer) is up */
+	(void)net_if_carrier_on(data->iface);
+}
+
+whd_result_t cy_buffer_pool_init(void *tx_packet_pool, void *rx_packet_pool)
+{
+	CY_UNUSED_PARAMETER(tx_packet_pool);
+	CY_UNUSED_PARAMETER(rx_packet_pool);
+
+	return 0;
+}
+
+static void scan_cb_search(whd_scan_result_t **result_ptr, void *user_data,
+			   whd_scan_status_t status)
+{
+	if (status == WHD_SCAN_ABORTED) {
+		k_sem_give(&airoc_wifi_data.sema_scan);
+		return;
+	}
+	if (status == WHD_SCAN_COMPLETED_SUCCESSFULLY) {
+		k_sem_give(&airoc_wifi_data.sema_scan);
+	} else if (status == WHD_SCAN_INCOMPLETE) {
+		if (user_data != NULL) {
+			if ((**result_ptr).SSID.length > 0) {
+				if (strcmp((**result_ptr).SSID.value,
+					   ((whd_scan_result_t *)user_data)->SSID.value) == 0) {
+					memcpy(user_data, *result_ptr, sizeof(whd_scan_result_t));
+				}
+			}
+		}
+	}
+}
+
+static void parse_scan_result(whd_scan_result_t *p_whd_result, struct wifi_scan_result *p_zy_result)
+{
+	if (p_whd_result->SSID.length != 0) {
+		p_zy_result->ssid_length = p_whd_result->SSID.length;
+		sprintf(p_zy_result->ssid, "%s", p_whd_result->SSID.value);
+		p_zy_result->channel = p_whd_result->channel;
+		p_zy_result->security = (p_whd_result->security == 0) ? WIFI_SECURITY_TYPE_NONE
+								      : WIFI_SECURITY_TYPE_PSK;
+		p_zy_result->rssi = (int8_t)p_whd_result->signal_strength;
+		p_zy_result->mac_length = 6;
+		memcpy(p_zy_result->mac, &p_whd_result->BSSID, 6);
+	}
+}
+
+static void scan_callback(whd_scan_result_t **result_ptr, void *user_data, whd_scan_status_t status)
+{
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)user_data;
+	whd_scan_result_t whd_scan_result;
+	struct wifi_scan_result zephyr_scan_result;
+
+	if (status == WHD_SCAN_COMPLETED_SUCCESSFULLY || status == WHD_SCAN_ABORTED) {
+		data->scan_rslt_cb(data->iface, 0, NULL);
+
+		return;
+	}
+	if ((result_ptr != NULL) || (*result_ptr != NULL)) {
+		memcpy(&whd_scan_result, *result_ptr, sizeof(whd_scan_result_t));
+		parse_scan_result(&whd_scan_result, &zephyr_scan_result);
+		data->scan_rslt_cb(data->iface, 0, &zephyr_scan_result);
```

Comment:
Depending on the context in which the callback is called, you might want call the callback from a separate context i.e., terminate the callstack here.

Reply:
Here (line#367) we process only received scan data. 

When we receive  status = WHD_SCAN_COMPLETED_SUCCESSFULLY or WHD_SCAN_ABORTED it means that all received scan data was processed so we notify  by calling callback (with NULL): data->scan_rslt_cb(data->iface, 0, NULL)

Reply:
Ah right, I should have posted this comment in line 361 where each result is conveyed up the stack.